### ISSUE TITLE
Unix domain socket support

### ISSUE THREAD

**Issue by tom-adsfund** [http-kit/http-kit#461]
Unix domain sockets can be connected to from NGINX. From [this example on Stackoverflow](https://stackoverflow.com/questions/58455529/how-do-i-connect-to-a-unix-domain-socket-running-an-http-server-using-netty) Netty allows using Unix domain sockets.

This could provide substantial performance improvements (from no TCP management).

**Comment by borkdude** [comment]
Maybe relevant, not sure: https://github.com/into-docker/unixsocket-http

**Comment by tom-adsfund** [comment]
@borkdude That would be a client-side. Although Netty can handle it, as in the example. (Big fan of bb btw!)

**Comment by borkdude** [comment]
@tom-adsfund http-kit isn't based on Netty. Aleph is though, which is a different project. And thanks! :)

**Comment by ptaoussanis** [comment]
@tom-adsfund Could you please clarify what the current status is here? As Michiel mentioned, http-kit does not use Netty - so I believe that support for domain sockets would need to be implemented some other way.

But before pursuing that idea further- we'd need to start by clarifying what the particular problem/s are that we'd be hoping to address with domain sockets.

Is there some reason to believe that performance is bottlenecked on the TCP connection between nginx and http-kit server? In my experience that would be highly unusual, and could indicate an nginx configuration issue.

**Comment by tom-adsfund** [comment]
@ptaoussanis It's really just based on the assumption that the OS would be able to manage things better using domain sockets and NGINX and http-kit would be able to multiplex over the domain socket (avoiding many file descriptors), rather than involving the whole TCP stack. Whether that applies in any specific case and implementation pairings etc, I don't know.

The exploration of whether this would work and how it compares I guessed would be under the scope of this library, based on its prior achievements in performance.

**Comment by ptaoussanis** [comment]
Okay- thanks for clarifying 👍 

So http-kit tends to strongly prioritise simplicity, which is both a source of its performance and (I believe) valuable for its own sake.

When it comes to performance, the key questions are:

1. How much of a benefit might some change provide (to what metrics, and under what conditions)?
2. What's the cost? (E.g. in terms of added complexity or other tradeoffs).

So in practice we'd need to start with at least a back-of-napkin estimate for [1] to know if it's worth investigating further.

At least for request->response time, my own intuition says that the nginx<->http-kit TCP comms is very unlikely to be a practical bottleneck since for Clojure applications it's almost always the application handlers that dominate.

There are other metrics that could be worth considering, and some scenarios might benefit more/less from something like domain sockets (e.g. maybe lots of long-held connections for a WebSocket server?).

Anyway, next steps if you or someone else would like to move this forward would be to more clearly define at least [1].

Hope that makes sense / seems reasonable!

Cheers :-)

**Comment by tom-adsfund** [comment]
@ptaoussanis In a quick search (really just finding SO answers), I found this:

https://redis.io/docs/reference/optimization/benchmarks/

"When the server and client benchmark programs run on the same box, both the TCP/IP loopback and unix domain sockets can be used. Depending on the platform, **unix domain sockets can achieve around 50% more throughput** than the TCP/IP loopback (on Linux for instance)."

https://github.com/goldsborough/ipc-bench

Basically 2x throughput at 1kb messages.

**Comment by ptaoussanis** [comment]
@tom-adsfund Hi Tom, thanks for that!

Unfortunately this information by itself isn't very useful. What matters is whether that improved throughput _at that point_ would make any significant difference to some metric/s we care about.

As a silly simplified example, imagine the numbers were something like this:
  - Time spent on http-kit servicing request: 10ms
  - Time spent on TCP proxying: 0.01ms
  - Total time spent: 10.01ms

Improving the TCP time by 50% (from 0.01ms to 0.005ms) would only impact the total response time by 0.005/10.01 ~0.5%.

We'd need to establish that the TCP time is actually contributing a non-trivial proportionate cost to some clearly defined target metric, in some realistic scenario.

It can be difficult to predict what the real effect of some change would have in a non-trivial context - esp. when we're talking about possible interactions between two pieces of software.

Would unfortunately really need a benchmark of some kind.

Hope that seems reasonable!

**Comment by ptaoussanis** [comment]
Neither the Redis nor IPC benchmarks are actually benching what matters here - which is the proportion of TCP throughput to total throughput when http-kit is involved, in a semi-realistic context. Do we know what proportion of time is spent by nginx:http-kit:application:TCP?

How do we know that http-kit won't bottleneck on other factors before any domain socket benefit can really be perceived? What's the currently handling time for the Webtorrent trackers?

I've seen surprisingly many cases where folks spent weeks or months trying to micro-optimize some obscure part of their tech stack without ever doing the profiling necessary to realise that they were spending 85% of their request handling time on some Ring middleware.

And relatively important details are easy to miss through overly-simple synthetic benchmarks. For example while a loopback Redis benchmark might indicate a significant benefit from domain sockets, in practice the difference to a real application throughput is likely to be affected by practical matters like:

- An application's Redis server is rarely on the same host as the application server.
- There's often more application servers than Redis servers, with connection pools and possible pipelining.
- The Redis client itself may involve overhead that.
- etc.

My point is just that if we're talking about a non-trivial cost to implement something, we'd need to establish that there'll be a realistic benefit of sufficient size to justify that cost. And I'm asserting that it's often difficult to predict the practical benefit of something like this without an attempt to quantify through at least semi-realistic measurement.

**Comment by tom-adsfund** [comment]
I agree that if it's costly to implement, there needs to be a clear isolated issue driving the request.

**Comment by kipz** [comment]
FWIW, I'm currently working in an environment where unix domain sockets are used as IPC between HTTP based web services running on the same machine. Given the co-location, I think the motivation for that is around security, as well as performance (avoiding the network stack).

I'm also keen on avoiding using any native bindings so that I can use graal `nativeimage` (more easily). Most web servers/clients in the JVM ecosystem are understandably trying to support many JDK versions, so tend to use JNI to achieve this. Also, they tend to use blocking IO. Anyway, I'm happy using JDK 17+ (I think this when they arrived - could be wrong), and UDS support is all nio based, so http-kit seemed like a good place to start.

As such, I've been playing with adding support over here:

https://github.com/kipz/http-kit/tree/kipz-unix-sockets

```clojure
(client/make-client {:address-finder (fn [uri]
                                                (UnixDomainSocketAddress/of "/var/run/backend.sock"))})
```
So far, this is just for the client, but works well with minimal changes (mostly changing `InetSocketAddress` to `SocketAddress`. Not sure if the server will be as easy. I'll report back.

I assume this project would be keen to maintain backwards compatibility with older JDKs, so maybe we'd need to plug in a JNI implementation for that (e.g. https://github.com/jnr/jnr-unixsocket)

**Comment by kipz** [comment]
FWIW, I've just updated my fork to support UDS on the server, and it was just as easy. Seems to work with graal `nativeimage` too:

```clojure
(server/run-server
   routes
   {:unix-socket-path "/var/run/backend.sock})
```

**Comment by tom-adsfund** [comment]
@kipz This is very cool. Is there a way you can make it obvious how people could use your version in their projects so that there could be feedback on whether this helps in other people's projects? This would add information so the people developing the main project can see whether it would be worth their time to merge. It's all pretty stable, so it's unlikely either version would diverge much over time.

**Comment by kipz** [comment]
@tom-adsfund what do you think is needed beyond what I've posted above? It's a case of building my fork and running:
`lein install`. I haven't published to clojars, but I could if anyone would find it helpful.

**Comment by tom-adsfund** [comment]
@kipz Just found this good article about this exact topic: https://lambdaisland.com/blog/2017-05-17-loading-clojure-libraries-directly-from-github

**Comment by ptaoussanis** [comment]
Closing since @kipz was able to provide a very clean PR (#510) with no obvious downsides.
PR is now merged on master, and on Clojars as `[http-kit "2.7.0-alpha4"]` 👍 

Note: this is an undocumented pre-release, and may be unstable or outright broken - please don't use for anything but curiosity/testing yet.

The UDS functionality is documented in the README, [here](https://github.com/http-kit/http-kit#unix-domain-sockets-udss).

Thanks a lot James! 💪

**Comment by tom-adsfund** [comment]
Excellent, great work everyone.
